File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int i = 0;
        int j = 0;
        int max = 0;
        while(j < s.length()){
            while(i < s.length() && set.contains(s.charAt(j))){
                set.remove(s.charAt(i));
                i++;
            }
            set.add(s.charAt(j));
            max = Math.max(max, j - i + 1);
            j++;
        }
        return max;

        
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hashset = set()
        result = 0
        i = 0
        j = 0
        while i < len(s):
            while s[i] in hashset:
                hashset.remove(s[j])
                j = j + 1
            hashset.add(s[i])
            result = max(result, len(hashset))
            i = i + 1
        return result
            



